### DISCUSSION TITLE
Allowing other domain types for FirstOrderFunction

### DISCUSSION THREAD

**Discussion by cpp977** [ceres-solver/ceres-solver#746] [Q&A] [answered]
Hello ceres-team!

The `FirstOrderFunction` functor has the follwing siognature:
```
  // cost is never null. gradient may be null. The return value
  // indicates whether the evaluation was successful or not.
  virtual bool Evaluate(const double* const parameters,
                        double* cost,
                        double* gradient) const = 0;
```
Is there a specific reason for only allowing `double` as the type for the domain of the cost function?
The codomain of the cost function needs to be real valued but why is the same assumed for the domain?

Thanks for answers!
Matthias

**Comment by alexsmac** [comment, marked as answer]
Are you asking whether the `parameters` could be another (real) floating point type (i.e. `float`) or whether it could be a complex type?

The answer to the latter is that Ceres has been developed for solving real-valued problems, and the algorithms would often require some modification to correctly support complex valued types (e.g. (L)-BFGS has complex variants).  Whilst that is doable, it would also have interface implications.  Whilst it would be possible to template the parameter type and only explicitly instantiate it for a subset of types, ensuring correctness for all would be a significant overhead that we have not had a use-case for.

**Reply by cpp977** [reply]
Yes I had complex types in mind. So allowing this would implicate quite some other changes.
Thanks for the quick answer.

**Reply by cpp977** [reply]
Do you have a suggestions for c++ libraries for complex-valued nonlinear optimization problems?

**Reply by sandwichmaker** [reply]
Sorry but I am not aware of complex valued optimization libraries.
